nl/nl_linux.go

```diff
@@ -238,54 +246,82 @@ func (req *NetlinkRequest) Execute(sockType int, resType uint16) ([][]byte, erro
 		defer s.Unlock()
 	}
 
-	if err := s.Send(req); err != nil {
-		return nil, err
+	sch := make(chan error, 1)
+
+	go func() {
+		sch <- s.Send(req)
+	}()
+
+	select {
+	case err := <-sch:
+		if err != nil {
+			return nil, err
+		}
+	case <-time.After(req.Timeout):
+		return nil, fmt.Errorf("netlink send timed out after %s", req.Timeout)
 	}
 
 	pid, err := s.GetPid()
 	if err != nil {
 		return nil, err
 	}
 
-	var res [][]byte
+	rch := make(chan error, 1)
 
-done:
-	for {
-		msgs, err := s.Receive()
-		if err != nil {
-			return nil, err
-		}
-		for _, m := range msgs {
-			if m.Header.Seq != req.Seq {
-				if sharedSocket {
-					continue
-				}
-				return nil, fmt.Errorf("Wrong Seq nr %d, expected %d", m.Header.Seq, req.Seq)
-			}
-			if m.Header.Pid != pid {
-				return nil, fmt.Errorf("Wrong pid %d, expected %d", m.Header.Pid, pid)
-			}
-			if m.Header.Type == syscall.NLMSG_DONE {
+	go func() {
+		var (
+			err  error
+			msgs []syscall.NetlinkMessage
+		)
+	done:
+		for {
+			msgs, err = s.Receive()
+			if err != nil {
 				break done
 			}
-			if m.Header.Type == syscall.NLMSG_ERROR {
-				native := NativeEndian()
-				error := int32(native.Uint32(m.Data[0:4]))
-				if error == 0 {
+			for _, m := range msgs {
+				if m.Header.Seq != req.Seq {
+					if sharedSocket {
+						continue
+					}
+					err = fmt.Errorf("Wrong Seq nr %d, expected %d", m.Header.Seq, req.Seq)
+					break done
+				}
+				if m.Header.Pid != pid {
+					err = fmt.Errorf("Wrong pid %d, expected %d", m.Header.Pid, pid)
+					break done
+
+				}
+				if m.Header.Type == syscall.NLMSG_DONE {
+					break done
+				}
+				if m.Header.Type == syscall.NLMSG_ERROR {
+					native := NativeEndian()
+					error := int32(native.Uint32(m.Data[0:4]))
+					if error == 0 {
+						break done
+					}
+					err = syscall.Errno(-error)
+					break done
+				}
+				if resType != 0 && m.Header.Type != resType {
+					continue
+				}
+				res = append(res, m.Data)
+				if m.Header.Flags&syscall.NLM_F_MULTI == 0 {
 					break done
 				}
-				return nil, syscall.Errno(-error)
-			}
-			if resType != 0 && m.Header.Type != resType {
-				continue
-			}
-			res = append(res, m.Data)
-			if m.Header.Flags&syscall.NLM_F_MULTI == 0 {
-				break done
 			}
 		}
+		rch <- err
+	}()
+
+	select {
+	case err := <-rch:
+		return res, err
+	case <-time.After(req.Timeout):
```

Comment:
I have seen reports where the hung happen in both the send and receive phase, I wanted the client to know during which phase the timeout happened.